### PR TITLE
[MU4] Midi Mapping. Configuration of midi devices

### PR DESCRIPTION
**Pull request by Eism**

### CODE REVIEW

**Comment by cbjeukendrup** [comment]
@Eism I implemented the `devicesChanged` notification for macOS using the CoreMIDI API. 
I think this way it is less power-consuming, because it makes use of the system notification instead of checking for devices every 500 ms.
You can find it here: https://github.com/musescore/MuseScore/commit/602260bd4afad3287ea95d0e617c3aac8613abeb

**Comment by Eism** [comment]
@cbjeukendrup I like your solution. I did not find such a solution for Linux and Windows. For macOS, I wasn't hoping to find such a solution after two fail :(

Ok, let's merge your solution with a separate PR. Need to test and remove extra code about my solution

**Comment by cbjeukendrup** [comment]
Good plan! I'll wait until this PR is merged, and then create a PR for the macOS solution.

**Review by RomanPudashkin — APPROVED** [review]

File: src/framework/midi/internal/midideviceslistener.cpp
Comment (igorkorsukov) [inline src/framework/midi/internal/midideviceslistener.cpp:53]: Would be nice add `th_` prefix

File: src/framework/midi/internal/midideviceslistener.cpp
Comment (igorkorsukov) [inline src/framework/midi/internal/midideviceslistener.cpp:58]: Would be nice add `th_` prefix

File: src/framework/midi/internal/midideviceslistener.cpp
Comment (igorkorsukov) [inline src/framework/midi/internal/midideviceslistener.cpp:69]: Would be nice add `th_` prefix

File: src/framework/midi/internal/midideviceslistener.cpp
Comment (igorkorsukov) [inline src/framework/midi/internal/midideviceslistener.cpp:60]: We need a flag like `m_running` (atomic), which must be set in false in the destructor, before the join, otherwise the join will never happen.

File: src/framework/midi/internal/platform/lin/alsamidiinport.cpp
Comment (igorkorsukov) [inline src/framework/midi/internal/platform/lin/alsamidiinport.cpp:76]: We need to add a mutex, now this method is called from two threads. Or you need to make sure that this method is not called from two threads

Comment (igorkorsukov) [reply src/framework/midi/internal/platform/lin/alsamidiinport.cpp:76]: Also for other platforms

**Comment by cbjeukendrup** [comment]
I am working on a follow-up for this PR, fixing the review issues, and I'll also fix a crash because m_actualDevicesCallback is called before it is set (`libc++abi: terminating with uncaught exception of type std::__1::bad_function_call: std::exception`).
Or is somebody else already doing this?

File: src/framework/midi/internal/midideviceslistener.cpp
Comment (cbjeukendrup) [reply src/framework/midi/internal/midideviceslistener.cpp:60]: Should `m_actualDevicesCallback` also be atomic? (Or I could just prohibit to set the callback while already running.)

**Comment by Eism** [comment]
@cbjeukendrup hmm
okay, if you have already started, then continue

File: src/framework/midi/internal/midideviceslistener.cpp
Comment (igorkorsukov) [reply src/framework/midi/internal/midideviceslistener.cpp:60]: m_actualDevicesCallback not posible to be atomic, so we should add mutex for it, or better yes, just prohibit to set the callback while already running.

**Pull request merged**
